Tangem/App/Services/TokenItemsRepository/CommonTokenItemsRepository.swift

```diff
@@ -7,273 +7,280 @@
+        lockQueue.sync {
+            migrateStorageIfNeeded()
+            updateCurrentStorageVersion()
+        }
     }
 
     deinit {
-        AppLog.shared.debug("TokenItemsRepository deinit")
+        AppLog.shared.debug("\(#function) \(objectDescription(self))")
+    }
+
+    private func migrateStorageIfNeeded() {
+        let migrator = StorageEntriesMigrator(
+            persistanceStorage: persistanceStorage,
+            cardID: key
+        ) { entries, cardID in
+            self.save(entries, to: \.entries, forCardID: cardID)
+        }
+        migrator.migrate(from: currentStorageVersion, to: actualStorageVersion)
+    }
+
+    /// - Warning: MUST BE called only AFTER the storage migration has been attempted,
+    /// otherwise user data may be lost.
+    private func updateCurrentStorageVersion() {
+        currentStorageVersion = actualStorageVersion
     }
 }
 
-// MARK: - TokenItemsRepository
+// MARK: - TokenItemsRepository protocol conformance
 
 extension CommonTokenItemsRepository: TokenItemsRepository {
-    var containsFile: Bool {
-        let entries: [StorageEntry]? = try? persistanceStorage.value(for: .wallets(cid: key))
-        return entries != nil
+    var isInitialized: Bool {
+        lockQueue.sync {
+            // Here it's necessary to distinguish between empty (`StorageEntry.V3.List.empty` value) and non-initialized
+            // (`nil` value) storage, therefore direct access to the underlying storage is used here
+            let list: StorageEntry.V3.List? = try? persistanceStorage.value(for: .wallets(cid: key))
+            return list != nil
+        }
+    }
+
+    var groupingOption: StorageEntry.V3.Grouping {
+        get {
+            return lockQueue.sync {
+                fetch().grouping
+            }
+        }
+        set {
+            lockQueue.sync {
+                save(newValue, to: \.grouping, forCardID: key)
+            }
+        }
+    }
+
+    var sortingOption: StorageEntry.V3.Sorting {
+        get {
+            return lockQueue.sync {
+                fetch().sorting
+            }
+        }
+        set {
+            lockQueue.sync {
+                save(newValue, to: \.sorting, forCardID: key)
+            }
+        }
     }
 
-    func update(_ entries: [StorageEntry]) {
+    func update(_ entries: [StorageEntry.V3.Entry]) {
         lockQueue.sync {
-            save(entries, for: key)
+            save(entries, to: \.entries, forCardID: key)
         }
     }
 
-    func append(_ entries: [StorageEntry]) {
+    func append(_ entries: [StorageEntry.V3.Entry]) {
         lockQueue.sync {
-            var items = fetch(for: key)
-            var hasAppended: Bool = false
+            var existingEntries = fetch().entries
+            var hasChanges = false
+            var existingBlockchainNetworksToUpdate: [StorageEntry.V3.BlockchainNetwork] = []
+
+            let existingEntriesWithIndicesGroupedByBlockchainNetworks = Dictionary(
```

Comment:
слишком много информации в названии. Для локальных функций думаю можно проще брать названия и сокращать какие-то части, т.к. их не надо запоминать для всего класса.
```suggestion
            let existingEntriesGroupedByNetwork = Dictionary(
```

Reply:
Поправил в тех местах, где еще актуально